src/librustc/ty/instance.rs

```diff
@@ -274,13 +274,29 @@ impl<'a, 'b, 'tcx> Instance<'tcx> {
                 _ => {
                     if Some(def_id) == tcx.lang_items().drop_in_place_fn() {
                         let ty = substs.type_at(0);
-                        if ty.needs_drop(tcx, ty::ParamEnv::reveal_all()) {
-                            debug!(" => nontrivial drop glue");
-                            ty::InstanceDef::DropGlue(def_id, Some(ty))
-                        } else {
-                            debug!(" => trivial drop glue");
-                            ty::InstanceDef::DropGlue(def_id, None)
-                        }
+                        let mut def = None;
+                        if let ty::Adt(inner_def, _) = ty.sty {
+                            let va_list_did = tcx.lang_items().va_list().unwrap();
+                            if inner_def.did == va_list_did {
+                                // Resolve `drop_in_place::<VaListImpl>` to `va_end`
```

Comment:
I couldn't add `lang = "va_end"` to `va_end` itself, so I did the next best thing: add a lang-item to the foreign module, and scan it every time to look for `va_end`.